### PR TITLE
Spoolman enhancements

### PR DESCRIPTION
**Pull request by CooperGerman**
According to discussion : #332 

Primary modifications are:
- `enable_remote_gate_map` added for `mmu_parameters.cfg` file
- `nb_gates` parameter added for `[mmu_server]` definition (This might be removed later as i understand that checks seem to be done on klippy side and not moonraker side).
- Added `SYNC` option for MMU_GATE_MAP which defaults to `enable_remote_gate_map` and can be used to persistently store the gate mapping into the spoolman db
- added `set/unset_spool`, `get_spool_info`, `remote_gate_map` methods
- the `remote_gate_map` method fetches spools assigned to current machine and maps them through the `MMU_GATE_MAP` command. This remote method is called in `MMU_BOOTUP_TASKS` when `enable_remote_gate_map` is set
- `!filament_used!` placeholder added (to be useful in `MMU_START_CHECK` or `MMU_START_SETUP` macros

### CODE REVIEW

**Comment by CooperGerman** [comment]
### Included in this PR
Main features of current version of this PR:
- `enable_remote_gate_map` added for `mmu_parameters.cfg` file
- `nb_gates` parameter added for `[mmu_server]` definition (This might be removed later as i understand that checks seem to be done on klippy side and not moonraker side).
- Added `SYNC` option for MMU_GATE_MAP which defaults to `enable_remote_gate_map` and can be used to persistently store the gate mapping into the spoolman db
- added `set/unset_spool`, `get_spool_info`, `remote_gate_map` methods
- the `remote_gate_map` method fetches spools assigned to current machine and maps them through the `MMU_GATE_MAP` command. This remote method is called in `MMU_BOOTUP_TASKS` when `enable_remote_gate_map` is set
- ~`!filament_used!` placeholder added (to be useful in `MMU_START_CHECK` or `MMU_START_SETUP` macros~
 _No need for this placeholder anymore, seeing as this might surely be done in OrcaSLicer (see this [PR](https://github.com/SoftFever/OrcaSlicer/pull/4771))_
- `!filament_names!` placeholder added (to be useful in `MMU_START_CHECK` or `MMU_START_SETUP` macros
- `automap_slicer_tools` added to enable mapping the slicer tools according to the filament names on the gates
- use `automap_slicer_tools` to auto remap in `MMU_START_SETUP` (This needs more testing as for now)


For now the `MMU_START_SETUP` command in the slicer needs to be updated as follows:
```
MMU_START_SETUP 
     INITIAL_TOOL={initial_tool} 
     REFERENCED_TOOLS=!referenced_tools! 
     TOOL_COLORS=!colors!  
     TOOL_TEMPS=!temperatures! 
     TOOL_MATERIALS=!materials! 
     PURGE_VOLUMES=!purge_volumes! 
     FILAMENT_USED=!filament_used! 
     FILAMENT_NAMES=!filament_names!
```
### Future works
- wait for review and discussion before updating doc
- better test implemented features (especially the prompt at auto remap)

**Comment by moggieuk** [comment]
This looks nice and I see the logic. I'm happy to incorporate this.  (I'm beginning to wish I saves the gate map as a dict rather than parallel arrays not because it is getting a bit unwieldy - but thats for another day).

I just push the v2.6.0 work into the "main" branch. Would you mind updating the PR for this to aid with merge.  The only conflict that is likely is in the "SAVE_VARIABLE" which I changed to enable batch updates to avoid TTC on some slow systems.

Small point, but I wonder if the new parameter should have "spoolman" in the name.  E.g. "enable_spoolman_ttg_remapping" or perhaps even make "enable_spoolman" an enumeration: 0 and 1 as current and add 2=auto sync ttg map with spoolman.  WDYT?

**Comment by CooperGerman** [comment]
> "enable_spoolman_ttg_remapping" or perhaps even make "enable_spoolman" an enumeration: 0 and 1 as current and add 2=auto sync ttg map with spoolman. WDYT?

I'm going for `enable_spoolman` : 
- `0` no spoolman integration (leaving as before)
- `1` spoolman integrated for color and material fetching (as before)
- `2` spoolman integrated with color and material fetching + store/retrieve gate_map from distant + auto remap 

I understant you would like to remove the auto `variable_automap_slicer_tools` from `[gcode_macro _MMU_SOFTWARE_VARS]` so that `enable_spoolman` configures all aspects ?

Furthermore, concerning the `nb_gates` parameters for the moonraker `[mmu_server]` declaration, should i remove it and fetch the nb_gates through remote_method calls from klippy ? or should i leave it and handle it to be generated through the shell installer ?

**Comment by moggieuk** [comment]
Why not just use `enable_spoolman`.  These are shipped complete macros so the average user doesn't need to fully understand.  Also, this means no need to run the `install.sh` upgrade user config files.  Just a simple update-manager refresh..

Didn't notice the nb_gates parameters for moonraker. Let me look again.

**Comment by moggieuk** [comment]
I think getting the number of gates can be done in a couple of ways:
The way I usually use is to look at the length of the TTG map (the KlipperScreen-Happy_Hare edition does that).
Or similarly you could query the configuration setting from klipper.

Maybe I'm missing something but why do you need the number of gates in moonraker?

**Comment by CooperGerman** [comment]
> I think getting the number of gates can be done in a couple of ways: The way I usually use is to look at the length of the TTG map (the KlipperScreen-Happy_Hare edition does that). Or similarly you could query the configuration setting from klipper.
> 
> Maybe I'm missing something but why do you need the number of gates in moonraker? BTW You have a lot of nice spoolman updates to moonraker. Have you asked for them to be merged?

I really don't need this information that much , i use it for some checks but more important, in order to generate the correct gate_map in moonraker myself. I think it will just be **given by klippy when calling the remote_method**. I can then check that no more than `nb_gates` spools have been set for the machine in spoolman db and other checks.

**Comment by CooperGerman** [comment]
> For the macro it might be more technically correct to retain the additional variable: `variable_automap_slicer_tools`.

I'm not sure i understand the need for this variable anymore, i was thinking about doing the following in the `MMU_START_SETUP` macro : 

```jinja
{% set automap_slicer_tools = 1 if printer.mmu.enable_spoolman > 1 else 0 %}
```
In this case i just need to add `enable_spoolman` to get_status in mmu.py. Maybe I'm not getting something here...

**Comment by CooperGerman** [comment]
> BTW You have a lot of nice spoolman updates to moonraker. Have you asked for them to be merged?

Concerning this, i moved the updates i had done in spoolman.py to mmu_server.py as it makes more sense for me. My changes where mainly about : 
- tracking gate assignments in spoolman (through a outdated method) which I ported to HH
- performing pre print checks (filament usage, material compatibility etc ...). Most of them are now done in HH, or already ported in spoolman, or will be available at slicer (at least soon for OrcaSlicer)
- dynamically remapping the gates (chich i ported to HH)

The only updates that might make sens to be made in spoolman could be the wrapper methods to set/unset a spool id, or display stuff in the console. 

The spoolman.py module is no longer being added through spoolman installation, but is now shipped by moonraker which makes the merge request more slow and hard to achieve. It looks like spoolman.py has been developped in a way to ease its utilization by third party modules as you did in mmu_server, this is way i moved my code ;)

**Comment by moggieuk** [comment]
> > For the macro it might be more technically correct to retain the additional variable: `variable_automap_slicer_tools`.
> 
> I'm not sure i understand the need for this variable anymore, i was thinking about doing the following in the `MMU_START_SETUP` macro :
> 
> ```
> {% set automap_slicer_tools = 1 if printer.mmu.enable_spoolman > 1 else 0 %}
> ```
> 
> In this case i just need to add `enable_spoolman` to get_status in mmu.py. Maybe I'm not getting something here...

`printer.mmu.enable_spoolman` (the true runtime setting / variable in the HH klipper module) isn't available as a printer variable.  You could use:
```yml
DEBUG_SEARCH_VARS S="enable_spoolman"
printer.configfile.config.mmu.enable_spoolman : 1
printer.configfile.settings.mmu.enable_spoolman : 1
```
Which 99.9% of the time would be correct.  The only time it wouldn't was if the user ran `MMU_TEST_CONFIG enable_spoolman=2" after booting.  This would dynamically change the value in HH but you would have no way to know.

To keep in simple, I'm fine with using the config value. I was just pointing out the very slight imperfection.

**Comment by moggieuk** [comment]
> The spoolman.py module is no longer being added through spoolman installation, but is now shipped by moonraker which makes the merge request more slow and hard to achieve. It looks like spoolman.py has been developed in a way to ease its utilization by third party modules as you did in mmu_server, this is way i moved my code ;)

Cool. Happy to be the "carrier" of this logic.

**Comment by CooperGerman** [comment]
Hi, i understand your point, it seems like adding `enable_spoolman` to the `get_status` dict of mmu.py makes it available to the printer as a runtime accurate variable isn't it ?

```py
LISTVARS SEARCH="enable_spoolman"

printer.configfile.config.mmu.enable_spoolman : 2
printer.configfile.settings.mmu.enable_spoolman : 2
printer.mmu.enable_spoolman : 2
```

 I think it might be a good idea to be able to modify it at runtime, i don't have the use case for now but I can imagine future features leveraging spoolman that might need to do it.

> I know the code is nearly complete but it would be helpful for a little doc to augment this wiki page: https://github.com/moggieuk/Happy-Hare/wiki/Spoolman-Support

I'm not so sure about how to edit the documentation directly on github as I don't have the permission (I think), it seems like i have to clone it locally, branch it at try to push a branch ? I would be glad to contribute to the doc :smiley:

**Comment by moggieuk** [comment]
Wiki:  Yeah, github doesn't make that easy.  You can clone locally, edit that one file then send to me.   I can fine tune the editing but it would be helpful to get description and use case from you.

Yes, we can add `enable_spoolman` to get_status().  I was trying to avoid growing that list but this is a case that won't change very often so adds no overhead.

**Comment by moggieuk** [comment]
Let me know when you are done and I will review and merge.

BTW I just pushed a few more updates. I don't think anything should mess with this merge.

**Comment by moggieuk** [comment]
Just checking in to see if you made any progress with this?  It's a good time for me to integrate and release before diving into supporting more MMU designs...
No pressure, just checking.

**Comment by CooperGerman** [comment]
Hi, i just came back to the project today, i was test running my updates and notices that i had some issues related to the purge matrix.
I indeed allow the gcode to have less than nb_gates tools used in it, thus the purge matrix might be 4 items instead of 9, 18, 81 expected... I was going to auto-generate missing points using `len(referenced_tools)` with `len(purge_matrix)` to determine the type of matrix and fill the "missing values in to 0.

I wanted to have your opinion on this matter , thanks in advance.

**Comment by CooperGerman** [comment]
I implemented it like [this](https://github.com/CooperGerman/Happy-Hare/commit/c2b47c620381d80c96c8e490d66e74b9f7e5bea0). In this case the MMU_START_SETUP needs to set it's PURGE_VOLUMES after the referenced tools have been assigned, is this problematic, it seems to be working as for now in my case (see [this](https://github.com/moggieuk/Happy-Hare/pull/335/commits/deca1506a86b9c901390d4dcc7ce9ed52df88f5a))

I've merged the main branch too without any problems, i'll be testing this more throughout the day !

**Comment by CooperGerman** [comment]
# PR development status

[ :heavy_check_mark:]  Add functionality to have spoolman remote db give the gate mapping and machine assignment 
- :heavy_plus_sign: push/pull capability with configurable source of truth

[ :heavy_check_mark:]  Add functionality to manipulate spoolman db from console interface in order to add/update/unassign gates and / or printer  
- :heavy_plus_sign: added (configurable) location field update
- :heavy_plus_sign: ability to edit other printers mapping 

[ :heavy_check_mark:]  Add functionality to be able to launch prints without having the entire gate map setup in the slicer
[ :heavy_check_mark:]  Add functionality to dynamically remap tools to gates according to the filament name 
- :heavy_plus_sign: added switches for spool_id (future use), material, color, closest color.
- :heavy_plus_sign: for closest color give a color matching status

[ :heavy_check_mark:]  Add documentation for all :heavy_check_mark: and :heavy_plus_sign: topics  
[ :no_entry_sign:]  Add functionality to check filament usage before print for each used tool in order to prevent running out.
_This should be done in the  [OrcaSlicer Soolman Integration PR](https://github.com/SoftFever/OrcaSlicer/pull/4771). Might actually be a good idea to keep this also at klipper side (to be confirmed)_

# FUTURE WORKS / LIMITATIONS / WEAKNESSES
- spool_id support to ensure once slicer give this option see [OrcaSlicer Soolman Integration PR](https://github.com/SoftFever/OrcaSlicer/pull/4771)
- Create endless spools when detecting multiple gates for one tool.
- Better testing _Given beta access will give us better coverage over various use cases_
- Better documentation _Might have to be appended with more details as users experience the features
- Some other stuff ?? @moggieuk

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:104]: Maybe rename to `extra_fields_available`

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:136]: Why no type hints?

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:280]: Type hints missing

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:272]: Type hints

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:319]: Stale

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:328]: Stale

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:336]: Stale

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:345]: Stale

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:363]: Stale

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:468]: Unused

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:539]: Type hint

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:583]: Return None hint

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:707]: Isn't that redundant, as the regexes are declared above and can't be dicts?

File: components/mmu_server.py
Comment (mawoka-myblock) [inline components/mmu_server.py:662]: Replace `processed by HappyHare` with global template string

File: config/base/mmu_parameters.cfg
Comment (mawoka-myblock) [inline config/base/mmu_parameters.cfg:326]: Use tabs instead of spaces

File: extras/mmu.py
Comment (mawoka-myblock) [inline extras/mmu.py:6455]: Maybe `if quiet: return`

File: extras/mmu.py
Comment (mawoka-myblock) [inline extras/mmu.py:7327]: `if quiet: return`

**Review by mawoka-myblock — CHANGES_REQUESTED** [review]
As I'm coming from nowhere and I've never coded for klipper nor Happy-Hare, I checked for Python errors/improvements I thought of, as that's something I can. Feel free to ignore this review completely if you don't find it useful! Really impressed by the work put into this PR!

**Comment by CooperGerman** [comment]
> As I'm coming from nowhere and I've never coded for klipper nor Happy-Hare, I checked for Python errors/improvements I thought of, as that's something I can. Feel free to ignore this review completely if you don't find it useful! Really impressed by the work put into this PR!

Thank you very much for the extra pair of eyes, always useful ! I'm taking a look and let @moggieuk be judge of the end result ;)

File: components/mmu_server.py
Comment (CooperGerman) [reply components/mmu_server.py:319]: i think this one has gone by now

Comment (mawoka-myblock) [reply components/mmu_server.py:319]: No, it hasn't. I reviewed minutes ago

**Comment by mawoka-myblock** [comment]
Sure! Just hoping I could assist, as I'm building my ercf and just searched for integrating it into spoolman and saw this PR!

File: components/mmu_server.py
Comment (mawoka-myblock) [reply components/mmu_server.py:707]: I'd probably remove that, add a comment instead like "Do it as above if so", as having less complexity is always better! But as I'm not the owner, @moggieuk can decide, that's only if this project was mine.

**Comment by CooperGerman** [comment]
> Sure! Just hoping I could assist, as I'm building my ercf and just searched for integrating it into spoolman and saw this PR!

Nice, hope this will be useful to you ! If you are interested the documentation associated with those changes has been published [here](https://github.com/CooperGerman/Happy-Hare/wiki) while in development you are welcome to take a look and hint us to missing information or unclear  explanations  =)

**Comment by mawoka-myblock** [comment]
Thanks! Don't have my parts yet and may report back in one week when I start getting into the software!

**Comment by moggieuk** [comment]
Ok. Merged with branch "v270" which I'll release soon as early access.  Once v270 branch is merged into main this PR will close.

**Pull request closed without merging**
